File: daily_problems/haskell/august_18.hs
-- This problem was asked by Facebook.
-- 
-- Given a list of integers, return the largest product that can be made by 
-- multiplying any three integers.
-- 
-- For example, if the list is [-10, -10, 5, 2], we should return 500, since 
-- that's -10 * -10 * 5.
-- 
-- You can assume the list has at least three integers.

import Control.Monad (filterM)

pset :: [a] -> [[a]]
pset = filterM (\_ -> [True, False])

maxProd :: (Num a, Ord a) => [a] -> a
maxProd s = maximum $ fmap (foldl1 (*)) [ s' | s' <- (pset s), (length s') == 3 ]

test1 :: IO ()
test1 = do
    let ts = [ -10, -10, 5, 2 ]
        res = maxProd ts
    putStrLn $ show res

main :: IO ()
main = test1

File: daily_problems/python/august/august_18.py
# This problem was asked by Facebook.
# 
# Given a list of integers, return the largest product that can be made by 
# multiplying any three integers.
# 
# For example, if the list is [-10, -10, 5, 2], we should return 500, since 
# that's -10 * -10 * 5.

from functools import reduce
import operator as op

def maxProd(s):
    p = reduce(op.mul, s[:3])
    for i in range(len(s) - 2):
        for j in range(i+1, len(s) - 1):
            for k in range(j+1, len(s)):
                if (s[i] * s[j] * s[k]) > p:
                    p = s[i] * s[j] * s[k]
    return p

def test1():
    ts = [ -10, -10, 5, 2 ]
    res = maxProd(ts)
    print(res)

def main():
    test1()

if __name__ == '__main__':
    main()
